### ISSUE TITLE
Product galleries and replacing PrettyPhoto

### ISSUE THREAD

**Issue by jameskoster** [woocommerce/woocommerce#10721]
**TL;DR** - Let's get rid of our lightbox script / option. Users can easily add one of the many plugins on .org for this functionality. Let's make the product image gallery a 'carousel' with touch navigation to vastly improve the experience on handheld devices.

Todo

- [x] Register but don't enqueue prettyPhoto and remove pP specific assets / markup relating to product galleries
- [x] Remove lightbox option
- [x] Refine markup (using `figure` tags) and update logic now that we don't need to account for lightbox galleries
- [x] Add FlexSlider assets
- [x] Style it all up real pretty like
- [x] Ensure `woocommerce_product_thumbnails_columns` filter still works as expected (adding the class as we do now should be enough)
- [x] Variation images
- [ ] Product Gallery Slider extension will probably have to be retired

====
We've recently spoken about replacing our lightbox script; PrettyPhoto. Whilst it has applications in extensions that will make it tricky to remove outright, it is overkill for our product gallery requirements. Worse still the handheld experience is very poor. Not only due to it's weight (~20kb not including the css) but mainly due to the overall experience. Here's what it looks like;

![prettyphoto](https://cldup.com/ttfG7qkPu5.gif)

To view the slideshow one has to click an image or a thumbnail. That opens the lightbox which displays the images in a container that is _smaller_ than the product page itself making it entirely redundant. To make matters worse, that smaller image is actually the _full size_ image uploaded by the store owner which could be huge and slow loading.

I think we should be aiming for something like this:

![new-gallery](https://cldup.com/sUBU3nF6pM.gif)

'Enlarging' the images is pointless unless it's via a pinching mechanism which isn't a particularly intuitive UX on web sites yet. Swiping through galleries seems like the best solution. It'd provide a familiar and intuitive experience for all smartphone users. Hiding the thumbs for the gallery (using the dot indicators instead) works well as it saves space and allows the main image to be center of attention whilst clearly indicating there are more to be viewed.

### Desktop

The desktop implementation is more open for discussion. Currently, clicking _any_ image in the gallery will open a lightbox. It works, but doesn't feel particularly intuitive to me. As the gallery thumbs are displayed a lot smaller than the main product image within the page I think there should be a different behaviour when clicking them.

What I'd like to see is that when a gallery thumbnail is clicked, the main product image is switched to the thumb you clicked. 

As for the lightbox itself, I think it's inclusion in core is debatable. Depending on the theme in use, it's not always necessary. I think this decision should be left to the end-user. We have an option to enable/disable the lightbox, but even when it's disabled the product images still link to the image directly which seems kinda pointless. I'd like us to remove the lightbox altogether. It'll make core a little lighter and there are loads of plugins on .org to add this functionality back in if folks need it. We can then remove the lightbox option as well.

### Potential replacement scripts

As for which script we might use to replace PrettyPhoto, that will need a little more research but https://github.com/jedrzejchalubek/Glide.js looks quite nice. Seems to be actively developed and is very light (~4kb before minification). It offers everything we need with one exception - no thumbnails. Perhaps we could contribute that feature to the project if we decide to use it.

**Comment by unfulvio** [comment]
check out also 
http://dimsemenov.com/plugins/magnific-popup/
github: https://github.com/dimsemenov/Magnific-Popup

**Comment by jameskoster** [comment]
@unfulvio 

> As for the lightbox itself, I think it's inclusion in core is debatable. Depending on the theme in use, it's not always necessary. I think this decision should be left to the end-user. We have an option to enable/disable the lightbox, but even when it's disabled the product images still link to the image directly which seems kinda pointless. I'd like us to remove the lightbox altogether. It'll make core a little lighter and there are loads of plugins on .org to add this functionality back in if folks need it. We can then remove the lightbox option as well.

xD

**Comment by unfulvio** [comment]
@jameskoster  yeah I saw that, but I suppose you wanted to debate this point thought and posted an alternative - I personally like it because it supports other content types which could be useful for other purposes such backend admin things - and it's 3kb

I know another lib that is just a carousel (no lightbox) https://github.com/kenwheeler/slick/ which I like but might be a bit heavier than Glide

**Comment by jameskoster** [comment]
Slick looks nice. RTL is a handy feature for sure.

**Comment by jameskoster** [comment]
Other potential scripts that include thumbnail navigation (but are a bit bigger)

* https://github.com/woothemes/FlexSlider
* https://github.com/nolimits4web/swiper/
* https://github.com/sachinchoolur/lightslider/

**Comment by BurlesonBrad** [comment]
https://github.com/RadGH/WooCommerce-Flickity-Gallery

**Comment by jameskoster** [comment]
Nice, Flickity looks like a winner.

**Comment by WPprodigy** [comment]
I've used slick in the past, but Flickity does look really promising. It seems well maintained, and from the same dev as isotope: http://isotope.metafizzy.co/

**Comment by jeffikus** [comment]
So, we have an existing product within Automattic that meets this need already -> **FlexSlider** (FS). I'm the lead on that and would be happy to contribute here in terms code and discussion around carousels/sliders. I've been pretty vocal and open about the fact that we do need more hands on the FS project, and including in WC core would certainly give it more exposure.

I can see the following benefits to using it:
(1) FS is already vastly used in most of Automattic and WooThemes themes, as well as a number of the most popular slider plugins in the community.
(2) to leverage WC dev community to get involved with FS,
(3) it shows a ​**massive** vote of confidence in our own slider library,

Regarding [RTL](https://github.com/woothemes/FlexSlider/pull/1374) - there is an open pull request for this for milestone 2.7.0 which is the next major release, but like I said, I can't maintain FS on my own unless I do it full time like I did last year. Having input from the WC community would be epic for the project!

I think it's a win win for both projects. 

Lastly, as an unbiased perspective, if you were going to choose one of the above that wasn't FS, I would suggest Swiper. iDangerous are pretty solid, they also make Framework 7 mobile dev framework, so they have extensive mobile experience, especially in terms of performance.

**Comment by BurlesonBrad** [comment]
# Hot Dang! 
Don't you kids ever get out and play?  Slap my fanny and call me Bertha ~  y'all know you work under the **same "roof"** (umbrella)? Hell, half the time I think **_I_** could tell all y'all more about the rest of your team than you could. :stuck_out_tongue_winking_eye: 

I reckon one of us ought to be a tacklin' dummy to test out Flickety on mobile.... Gimme a bit.  Here's the mini version of @jeffikus Flexin' 

![screenshot_2016-04-26-14-38-44](https://cloud.githubusercontent.com/assets/6208757/14831951/6fdc1804-0bbd-11e6-8fd9-832d0070aba8.png)


## _y'all still crack me up sometimes_

**Comment by BurlesonBrad** [comment]
Ok, Here y'all go.  I'm gonna go ahead and tag y'all so you can give the Flickety a spin real quick and see what it looks like (on Galleria) 
@jeffikus @jameskoster @WPprodigy @unfulvio @mattyza @mikejolley 
### [https://bradgriffin.me/product/awesome-product-title/](https://bradgriffin.me/product/awesome-product-title/)

![screenshot_2016-04-26-15-54-12](https://cloud.githubusercontent.com/assets/6208757/14833996/9921ab7a-0bc7-11e6-990e-76a0c221dc31.png)

(sorry... guess I ought to tag @slash1andy ) :wink: 

Looks like there's some spacing issues but........

**Comment by jameskoster** [comment]
I did already list Flex as a possibility @BurlesonBrad xD Of course it's my preference that we use our own script, but it has to be fit for purpose.

It does everything we need except rtl, but that's coming soon. When is 2.7 due @jeffikus ?

**Comment by jeffikus** [comment]
@jameskoster 2.7.0 currently has no due date as I'm wrapping up 2.6.1 - but 2.7.0 is currently 17% complete - I could shift priorities around if necessary though? https://github.com/woothemes/FlexSlider/milestones/2.7.0

**Comment by thomasguillot** [comment]
@jeffikus I'll give you a hand with 2.7.0 next week

**Comment by jeffikus** [comment]
Update for you on 2.6.1 - it's nearing completion and will be out next week. There are 4 projects in the works for FlexSlider:
1 - 2.6.1 release
2 - 2.7.0 release
3 - Update the demos (there are a few issues and changes that need to be made)
4 - Optimize theme/plugin dependancy for maintenance and for projects such as WooSlider to use FlexSlider.

**Comment by SteveHoneyNZ** [comment]
I understand the internal preference for flexslider but personal vote would be for [slick](http://kenwheeler.github.io/slick/) also.

I spent a long time experimenting with carousel scripts last year and it stands out a mile in my experience. Personal favourite features:

- Breakpoints for settings
- Lazyload with progressive/ondemand options (granted not best way to implement it but handy at times)
- customPaging function works really well, highly customisable
- rows option for grid layout
- touchThreshold is a nice thing to have control over
- Using CSS transform option and setting cssEase with a cubic-bezier value results in best transitions I've seen across devices. [Handy tool for experimenting](https://matthewlein.com/ceaser/)
- Using the data-slick attribute to pass in options is very useful in some contexts

There's also already a solid add on for a [lightbox](https://github.com/mreq/slick-lightbox)

Just my own thoughts but seemed relevant and figure another viewpoint doesn't hurt when it's not a list of demands ;)

@kenwheeler @mreq

**Comment by kenwheeler** [comment]
I personally vote for slick :) . That said I think slick or flickity would both work fine. The main difference being licensing. Slick has some great features that are handy for prod level work. Flickity is super well written and takes a more physics based approach. I would generally recommend examining handfeel on each and taking it from there.

**Comment by mreq** [comment]
slick's great ;)

**Comment by jeffikus** [comment]
FS 2.6.1 Tagged -> https://github.com/woothemes/FlexSlider/releases/tag/2.6.1

**Comment by jeffikus** [comment]
@jameskoster any updates on this? the dev team talked about this again? I'm working on 2.7.0 some more next week, so would be good to know.

**Comment by jameskoster** [comment]
@jeffikus the plan is for this to go in 2.7 which is a while away yet. We haven't set a deadline to decide on which script to use. Slick and Flickety are getting a lot of love here though so Flex will need to match up with them - particularly with rtl support.

Any input here @mikejolley @claudiosmweb ?

**Comment by claudiosmweb** [comment]
@jameskoster @jeffikus I love FlexSlider, It was the first thing made by WooThemes that I have ever used... But yes, we need RTL support.
I prefer use FlexSlider of course, since we control it and we can make sure that is always on development.

I understand and agree on the Lightbox issue, still I think that we need something to display more details of an image... I think it would be interesting to use along with the slideshow an image zoom script too, like this one: http://www.jacklmoore.com/zoom/

**Comment by jameskoster** [comment]
That'd mean we'd have to retire our Zoom extension? :p I prefer zooming to popups but feel like it's better suited to plugins. A lot of stores don't have particularly high quality photos or even need this feature. We can talk about it more on WooTrip :)

**Comment by emzo** [comment]
Here's a really nice lightweight carousel from the guys at Filament Group with snap point support. It's called Snapper.

https://www.filamentgroup.com/lab/introducing-snapper.html
http://filamentgroup.github.io/snapper/demo/
https://github.com/filamentgroup/snapper

It's being used on the [Vineyard Vines](http://www.vineyardvines.com/womens-tops/mixed-media-sleeveless-top/2K0779.html) store.

**Comment by jeffikus** [comment]
@jameskoster @claudiosmweb would love to sit and chat more about this with you guys next week so I can line things up - I have some active RTL pull requests for the 2.7 Flexslider milestone, so I can prioritize those.

**Comment by rgb4u** [comment]
Hi, i don't know if its already exist, or not, but it will be nice to have automatic
 image interchange for product thumbnail and images in gallery
 using Foundation 6 Interchange, along with this tutorial:
https://www.jaredcobb.com/automatically-get-responsive-images-with-foundation-interchange-in-wordpress/

So we/our clients can just upload product photo, and images in gallery, and on each
media query breakpoint our images will fit for this particular breakpoint only. 
Sorry for my English.

**Comment by jeffikus** [comment]
Update on Flexslider RTL - I've merged a PR into the 2.7.0 release branch https://github.com/woothemes/FlexSlider/pull/1561 

Any testing feedback would be appreciated!

**Comment by jameskoster** [comment]
@jeffikus any ETA for 2.7?

**Comment by jeffikus** [comment]
@jameskoster I can deploy it with RTL support soon if you need it? I've just let it run free for a bit so the community can test. 

I'm planning on doing a last call for testing on Monday and then I'm going to bump all the remaining issues (that I don't get time to fix next week) to the next release, if you are happy with that?

**Comment by jameskoster** [comment]
No rush. I'm hoping to work on this sometime this week. You can wait until then if you like. I'll let you know if I run in to any major issues.

**Comment by jameskoster** [comment]
Just needs someone to work on variation images and it's done.
